### DISCUSSION TITLE
Release 3.1

### DISCUSSION THREAD

**Discussion by spamatica** [muse-sequencer/muse#1133] [Team Posts]
So, it would be nice if we could make the final 3.1 release before the new year. 
What we have now seems pretty stable (always a list of nagging bugs of course)
Mostly looking at you Tim, did you have any bigger changes that you really wanted to add before a release?

**Comment by terminator356** [comment]
Thanks for asking. I noticed you mentioned a pre4? Fine with me if you like.

**I have a request:** Would you guys be terribly put off if I merged my audio_convert branch?
**Please please, I must dump it into master before I go nuts !!!** I really need to work on it in master.
I just finished a tough merge of master into it after a 12 month absence, and the merge
 before that was also very tough. Took me two days. I have fixed a remaining issue and it works well now.

I solemnly promise to make it as unobtrusive as possible.
It does not run the converters unless required, ie. unless someone uses it.
I promise to address any issues that might come up when it is **not** in use.
**If I think it won't work, I won't merge.**

There are three very tough very important things to have in a DAW:
1) Latency correction.
2) Automatic 'live' non-destructive wave re-sampling (load any wave anytime).
3) Time stretching. 

1 Got it. Check.
2 and 3 That's what I've got here in this branch.
All three have been my pride and joy, my puppies for so long.
For about 8 years 2 and 3 have sat in this branch.
I can't take the master merge updates any more !!!

The branch is a commented mess (of course).
It'll take me a day just to clean it up.

**I leave you with this thought:** Once this whole deal is done, if we can do it well, MusE 
can finally take one more giant leap:
our wave parts can finally behave exactly like our midi parts. They will stay in place
 when you change tempo and will automatically stretch just like midi parts.
This is a very desirable goal and makes MusE much more logical in usage.
I will not stop until we have... MusE-o-Dyne !
[https://www.celemony.com/en/start](url)

**Comment by kybos** [comment]
Do as you wish, I am not (yet) expert enough in MusE core to be able to judge ;-).

**Comment by kybos** [comment]
As to release before new year:
I won't be available between 20.12.2019 and 6.1.2020, so I won't be able to help, should that become necessary. Probably not relevant, I did not make any real critical developments in MusE, but just for the record...

**Comment by spamatica** [comment]
I guess it's not optimal to merge work in progress but I value your sanity
too so I think you should merge it. :)
Getting it out will get exposure from more testers too. And who am I to
judge, all I ever do feels like work in progress :P

I can try to make a pre4 release today and you can merge it after that? It
would be good if you could limit the comment-mess you mentioned though..


Den tis 10 dec. 2019 kl 06:40 skrev kybos <notifications@github.com>:

**Comment by terminator356** [comment]
Deal. I'll wait. Thanks. Whenever you're ready, your call. Take your time.

**Comment by spamatica** [comment]
Hit a speed bump.
Building clean I get an error that ui_midifilter.h is missing while
building components. It is generated in mplugins so I moved that before in
the CMake file but then I get an error that ui_itransformbase.h is missing,
it seems to be generated in components, so, circular reference. One of them
should be moved I guess?


Den tis 10 dec. 2019 kl 17:38 skrev Tim <notifications@github.com>:

**Comment by terminator356** [comment]
Hi, I'll try moving them now...
transformbase ui really should be in mplugins folder, and anything else that belongs there.

My recent move of Appearance source files into components to be together with its ui file there,
 seemed to have worked well.

**Comment by terminator356** [comment]
OK I moved several things around.
All source files are together with their UI files now.
There is no more spreading around of modules.
Also cliplist was not in the cmake includes.
I think I got everything.
Hopefully this ends this specific type of mysterious cyclic dependency problem,
 that even some users have reported. (The guy with -j2 vs. -j4 which we closed.)

**Comment by spamatica** [comment]
I actually had to do some more shuffling of files. See the last commit.
Very strange. I think it did compile when I built it from the repository
but the version that was created by the packaging mechanism was somehow
different and did not compile without these changes.
It's late so I won't complete the process (nothing is public yet) but it's
all prepared so committing to master is fine if anyone wants to do that.


Den tis 10 dec. 2019 kl 22:47 skrev Tim <notifications@github.com>:

**Comment by terminator356** [comment]
**Wait!** I just fixed three mixer problems, one of them a long-time pest!
    - FIXED: Mixer strips missing if loading an older song saved before 20.10.2019
    - FIXED: Duplicate strips found in songs saved before 20.10.2019
    - FIXED: Long-time mixer problem with restoring size.

The first two hopefully fix your recent issue report, I'll answer it later.

The first  was an obvious bug, I forgot to reset a loop iterator inside another loop.
Resulting in only a couple of strips where there should have been many.

The second I observed in my songs saved before that time.
The mixers today, would try to create strips for the duplicates.
Likely flaws in old saving code?

That last one drove me nuts for years.  I tried several fixes.
You would save a dozen-track song with the mixer open wide, and when you
reopen the song the mixer is skinny again. Frustrating.
I am pretty certain I got it this time.
Here's my in-code comment from app.cpp : MusE::showMixer1 (or 2):
>  // NOTICE!!! Our mixers set their own maiximum size according to their content.
>  //           Therefore if the mixer is ALREADY OPEN, it will have a maximum size
>  //            imposed on it, which may be SMALLER than any new size we might
>  //            try to set after this. Therefore we MUST RESET the maximium size now,
>  //            BEFORE any attempts to set size. As per docs:
>  mixer1->setMaximumSize(QWIDGETSIZE_MAX, QWIDGETSIZE_MAX);

Abut your move: Oh Geez! Freakin' brilliant catch bud! Sorry about that.
Those miditransform source files were originally in muse not mplugins since 
 that module is technically not a 'midi input plugin' like the others found in mplugin.
However, my mistake was throwing them into components.
Because their 'order of found usage' and creation is the same as those plugins, in app.cpp.
So good catch I think. Probably a good place to put that module anyway.
Hopefully puts this long episode to rest...

**Comment by terminator356** [comment]
Update: And as is typical, a rewrite of that last (third) fix. Tested OK.

**Comment by spamatica** [comment]
Alright, I'll redo the pre4 package tonight then :)


Den ons 11 dec. 2019 kl 07:10 skrev Tim <notifications@github.com>:
